### PR TITLE
CODEOWNER resolving utils

### PR DESCRIPTION
**Pull request by zkdev**
**What this PR does / why we need it**:
Add utils to resolve codeowners not only by email, but an extensible set of attributes.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```other operator

```

### CODE REVIEW

File: github/codeowners.py [hunk 1]
```diff
@@ -84,6 +88,107 @@ def _first(iterable):
...
 
 
+@dataclasses.dataclass(frozen=True)
+class CodeOwnerGithubUser:
+    user: str
+    source: str
+
+
+@dataclasses.dataclass(frozen=True)
+class CodeOwnerPersonalName:
```
Comment (ccwienk) [inline github/codeowners.py:98]: nit: it is just a PersonalName

File: github/codeowners.py [hunk 2]
```diff
@@ -84,6 +88,107 @@ def _first(iterable):
         return None
 
 
+@dataclasses.dataclass(frozen=True)
+class CodeOwnerGithubUser:
```
Comment (ccwienk) [inline github/codeowners.py:92]: GithubUser

File: github/codeowners.py [hunk 3]
```diff
@@ -84,6 +88,107 @@ def _first(iterable):
...
+
+@dataclasses.dataclass(frozen=True)
+class CodeOwnerPersonalName:
+    firstName: str
+    lastName: str
+    source: str
+
+
+@dataclasses.dataclass(frozen=True)
+class CodeOwnerEmail:
```
Comment (ccwienk) [inline github/codeowners.py:105]: EmailAddress

File: github/codeowners.py [hunk 4]
```diff
@@ -84,6 +88,107 @@ def _first(iterable):
...
+    email: str
+    source: str
+
```
Comment (ccwienk) [inline github/codeowners.py:108]: define a base class `UserInfo(Base?)` that defines the `source` attr, which is common to all DCs you defined in this diff

File: github/codeowners.py [hunk 5]
```diff
@@ -84,6 +88,107 @@ def _first(iterable):
...
+
+@dataclasses.dataclass
+class CodeOwner:
```
Comment (ccwienk) [inline github/codeowners.py:111]: that's not exactly how I imagined this :-/

also, I think this code actually belongs into the delivery-service repository

I would like to have a more abstract object representing a UserIdentity (maybe call it like this?)

A UserIdentity consists of a non-empty collection of (typed) UserInfo objects:
- email-address
- github-user
- personal-name
- ... (for now, there are no more, but there might be. e.g. "d-user")

File: github/codeowners.py [hunk 6]
```diff
@@ -84,6 +88,107 @@ def _first(iterable):
...
+    github: typing.Optional[CodeOwnerGithubUser]
+    personalName: typing.Optional[CodeOwnerPersonalName]
+    email: typing.Optional[CodeOwnerEmail]
+
+    @staticmethod
+    def create(
```
Comment (ccwienk) [inline github/codeowners.py:117]: as mentioned above, this should be a sequence..

File: github/codeowners.py [hunk 7]
```diff
@@ -84,6 +88,107 @@ def _first(iterable):
...
+    PERSONAL_NAME = 'personalName'
+    GITHUB_USER = 'githubUser'
+    EMAIL = 'email'
+    UNKNOWN = 'unknown'
+
+    @staticmethod
+    def for_codeowner_attribute(
+        attribute: typing.Union[CodeOwnerGithubUser, CodeOwnerEmail, CodeOwnerPersonalName],
+    ) -> 'CodeOwnerMetadataTypes':
+        if isinstance(attribute, CodeOwnerGithubUser):
```
Comment (ccwienk) [inline github/codeowners.py:171]: to convert into enums, simply use the enum's c'tor

--> ```
CodeOwnerMetadataTypes('personalName') is CodeOwnerMetadataTypes(CodeOwnerMetadataTypes.PERSONAL_NAME)
```

also, by convention (in this repository), use singular for enum-type-names, please

File: github/codeowners.py [hunk 8]
```diff
@@ -84,6 +88,107 @@ def _first(iterable):
...
+        elif isinstance(attribute, CodeOwnerEmail):
+            return CodeOwnerMetadataTypes.EMAIL
+        elif isinstance(attribute, CodeOwnerPersonalName):
+            return CodeOwnerMetadataTypes.PERSONAL_NAME
+        else:
+            return CodeOwnerMetadataTypes.UNKNOWN
+
+
+def iter_codeowner_attribute_dict(
+    codeowner: CodeOwner,
```
Comment (ccwienk) [inline github/codeowners.py:182]: not needed, i think

**Review by ccwienk — CHANGES_REQUESTED** [review]

**Comment by zkdev** [comment]
https://github.com/gardener/cc-utils/pull/710

**Pull request closed without merging**